jdb/assignment2/tree.py

```diff
@@ -0,0 +1,196 @@
+typing, and types complement unit tests:
+1. they help the readers of the code,
+2. they help the author to design a good consistent API,
+3. type checkers find errors statically, thus saving some class of errors
+   before the code is sent to production.
+4. static analysis helps with refactoring, and refactoring is a fact of life for
+    medium to big sized project.
+We suggest and warmly support the use of Python type annotations :)
+More info at: https://docs.python.org/3/library/typing.html
+
+To run the mypy static type checker (http://mypy-lang.org/):
+$ mypy tree.py
+"""
+
+import unittest
+from typing import Generator, Tuple, TypeVar
+
+T = TypeVar('T')  # the generic type of values held in each node.
+
+class Node:
+  def __init__(self, value : T,
+               left: Tuple = None,
+               right: Tuple = None,
+               parent: 'Node' = None,
+               depth: int = 0):
+
+    self.value = value
+    self.parent = parent
+    self.depth = depth
+    self.left = Node(*left, parent=self, depth=depth+1) if left else None
+    self.right = Node(*right, parent=self, depth=depth+1) if right else None
+
+  def _ancestors(self) -> Generator['Node', None, None]:
+    """Yields the current node, then the ancestor nodes up to the root.
+
+    The first element yield is the node itself (in this sense, the node is
+    considered an ancestor of itself).
+    """
+    yield self
+    if self.parent:
+      yield from self.parent._ancestors()
+
+  def print_ancestors(self):
+    print(' '.join(str(i) for i in self._ancestors()))
+
+  def lowest_common_ancestor(self, node1: 'Node', node2: 'Node') -> 'Node':
+    """Returns the node that is the lowest common ancestor of node1, node2."""
+    deeper, shallower  = ((node2, node1)
+                          if node2.depth > node1.depth else (node1, node2))
+    for _ in range(deeper.depth - shallower.depth):
+      deeper = deeper.parent
+
+    # Note: because _ancestors() returns an iterator, and the iterator behaves
+    # lazily, ancestors() will not reach to the top of the tree, unless
+    # required. Each iterator will be pulled just enough times to find the
+    # common ancestor.
+    for a1, a2 in zip(deeper._ancestors(), shallower._ancestors()):
+      if a1 == a2:
+        return a1
+    raise ValueError('The two nodes have no common ancestor, '
+                     'Did you check the graph is actually connected?')
+
+
+def search(root: Node, value: T):
+  """Helper function handy for the unit tests: find the node with 'value'."""
+  for subtree in root.left, root.right:
+    if subtree:
+      try:
+        return search(subtree, value)
+      except KeyError:
+        pass
+  if root.value == value:
+    return root
+  raise KeyError("Value '%s' not found in the tree." % value)
+
+
+def check_loops(value: T,
+                left: tuple = None,
+                right: tuple = None,
+                seen: set =None):
```

Comment:
In my opinion, the interface of this function could be more convenient if if could take a Node as an input.

For example, I don't know how to test the following one-node graph with this function:

```
tricky_node = Node("value")
tricky_node.left = tricky_node
tricky_node.right = tricky_node
tricky_node.parent = tricky_node
# how to use check_loops?..
```

Reply:
The idea was not to check a node, but to check the input for the constructor of Node (the method Node.__init__). So it does not take a Node as an input but the recursive tuple of tuples data structure that you can find below in the unit test. The idea is that the user checks the input before creating a Node, we expect Node to only be created via the constructor and never mutated after. If this is true Nodes, can never be incorrect.

This is good feedback, I agree the current code as you show in your example allow to create bad nodes. Not only there is a loop, in the example you show, but also, the height is incorrect!

If you want check loops to work on Nodes, here is such a method:

class Node:
  def check_loops(self, seen=None):
    "Returns None if all went well, or raise a ValueError if a loop is detected.."
    if seen is None:
        seen = set()
    if id(self) in seen:
      raise ValueError("Loop detected")
    else:
      seen.add(id(self))
    if self.left:
        self.left.check_loop()
    if self.right:
        self.right.check_loop()

I had a first version that was integrating the loop check in the Node constructor, but the code started to be a bit complicated, messy. I found this version more readable: after all, the "interview question"  does not request checking for loop: so better solve the exact problem first cleanly, then improve on the solution if there is still time.